### ISSUE TITLE
Workgraph test hangs with gcc 6.1 

### ISSUE THREAD

**Issue by dsunder** [kokkos/kokkos#988]
Workgraph_fib hangs with gcc 6.1 on kokkos-dev

**Comment by ibaned** [comment]
#989 disabled this test

**Comment by ibaned** [comment]
Running under GDB. 10 threads. All of them are inside `pop_work`, which is already a bad sign. 9 of them are waiting on `atomic_compare_exchange` of the queue range. The 10th is doing the "guess" operation that one more work item will be added. Note that this guess is only done if there is no work in the queue... so current best guess as to the situation, there is no work in the queue and everyone is waiting for more...

**Comment by ibaned** [comment]
At the time of livelock, `range.first` and `range.second` are equal, and three or four less than `m_total_work`. Work items are somehow slipping through the cracks.

**Comment by ibaned** [comment]
Yep, there is somehow one item that gets pushed but never popped. We must have a bug somewhere where our begin/end indices are able to skip over a queue entry without it getting properly popped.
```
tid 2 w(107, 107)
item 4 pushed by 4 to queue loc 105 but never popped
m_queue(105) was popped by -1, at the time had -1, now has 4
m_queue(104) was popped by 8, at the time had 2, now has 2
```
latest debug prints are kept in the `issue-988` branch

**Comment by crtrott** [comment]
That is gonna be fun to find, enjoy ...

**Comment by ibaned** [comment]
Fixed yesterday with help from @hcedwar .
Basically, it looks like a loop that is just repeatedly doing a volatile read hangs waiting for the new value to appear even though another thread has done a volatile write to that location. Replacing the volatile read and write with `atomic_exchange` seems to fix the hang. 
This fix is in PR #1107.
